pwiz_tools/Skyline/Controls/Graphs/GraphSpectrum.cs

```diff
@@ -911,6 +987,240 @@ private void ClearGraphPane()
+                                if (spectrumInfo.IsBest)
+                                {
+                                    var iBest = precursorSpectra[i].Spectra.IndexOf(s =>
+                                        Equals(s.Name, spectrumInfo.Name) &&
+                                        Equals(s.LabelType, spectrumInfo.LabelType));
+                                    if (iBest != -1)
+                                    {
+                                        precursorSpectra[i].Spectra[iBest] = new SpectrumDisplayInfo(
+                                            precursorSpectra[i].Spectra[iBest].SpectrumInfo, precursor.Precursor,
+                                            replicateName,
+                                            matchingFile.FilePath, 0, spectrumInfo.RetentionTime, true);
+                                    }
+                                }
+
+                                if (!dictReplicateNameFiles.TryGetValue(replicateName, out var setFiles))
+                                {
+                                    setFiles = new HashSet<string>();
+                                    dictReplicateNameFiles.Add(replicateName, setFiles);
+                                }
+
+                                setFiles.Add(spectrumInfo.FilePath);
+                            }
+
+                            // Determine if replicate name is sufficient to uniquely identify the file
+                            foreach (var spectrumInfo in spectraRedundant)
+                            {
+                                var replicateName = spectrumInfo.ReplicateName;
+                                if (replicateName != null && dictReplicateNameFiles[replicateName].Count < 2)
+                                    spectrumInfo.IsReplicateUnique = true;
+                            }
+                            spectraRedundant.Sort();
+                            precursorSpectra[i].Spectra.AddRange(spectraRedundant);
+                        }
+                    }
+                    catch (Exception)
+                    {
+                        // ignored
+                    }
+                }
+
+                _spectraLoadCache = Cache.Create(null, precursorSpectra, settings);
+            }
+
+            public bool GetPrecursorStrings(IReadOnlyList<PrecursorSpectra> spectra, out string[] outStrings)
+            {
+                if (Cache.GetValue(_precursorStringCache, out var value, spectra))
+                {
+                    outStrings = (string[])value;
+                    return true;
+                }
+                outStrings = (spectra ?? Array.Empty<PrecursorSpectra>()).Select(s => s.PrecursorString).ToArray();
+                _precursorStringCache = Cache.Create(outStrings, spectra);
+                return false;
+            }
+
+            public void Dispose()
+            {
+                _timer.Dispose();
+            }
+
+            private class Cache
```

Comment:
The "Cache" class here seems to make things more confusing, since it ends up obscuring the type of the keys and values that are used here.